state/executor.go

```diff
@@ -471,25 +500,28 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, er
 
 	var result *runtime.ExecutionResult
 	if msg.IsContractCreation() {
-		result = t.Create2(msg.From, msg.Input, value, gasLeft)
+		result = t.Create2(msg.From, msg.Input, value, availableGas)
 	} else {
 		txn.IncrNonce(msg.From)
-		result = t.Call2(msg.From, *msg.To, msg.Input, value, gasLeft)
+		result = t.Call2(msg.From, *msg.To, msg.Input, value, availableGas)
 	}
 
-	refund := txn.GetRefund()
-	result.UpdateGasUsed(msg.Gas, refund)
+	// TODO: remove knowledge of TxTypeState
+	if msg.Type != types.TxTypeState {
+		refund := txn.GetRefund()
+		result.UpdateGasUsed(msg.Gas, refund)
 
-	// refund the sender
-	remaining := new(big.Int).Mul(new(big.Int).SetUint64(result.GasLeft), gasPrice)
-	txn.AddBalance(msg.From, remaining)
+		// refund the sender
+		remaining := new(big.Int).Mul(new(big.Int).SetUint64(result.GasLeft), gasPrice)
+		txn.AddBalance(msg.From, remaining)
 
-	// pay the coinbase
-	coinbaseFee := new(big.Int).Mul(new(big.Int).SetUint64(result.GasUsed), gasPrice)
-	txn.AddBalance(t.ctx.Coinbase, coinbaseFee)
+		// pay the coinbase
+		coinbaseFee := new(big.Int).Mul(new(big.Int).SetUint64(result.GasUsed), gasPrice)
+		txn.AddBalance(t.ctx.Coinbase, coinbaseFee)
 
-	// return gas to the pool
-	t.addGasPool(result.GasLeft)
+		// return gas to the pool
+		t.addGasPool(result.GasLeft)
+	}
```

Comment:
What do you think about extracting the `apply` method to be reimplemented by regular transactions and these TxTypeState transactions?

That way we can avoid this if spaghetti

Reply:
Separated eddbd35203739b6fe17c9e13481e676cf9fa36d4